### PR TITLE
Revamp Admin Module

### PR DESCRIPTION
**Pull request by chaveiro**
Fixs or add the follow:
-  Full components manager
-- Add component uninstall
-- Add component install
-- Add components available list
-- Notify if updates available
- added disks read load statistic
- improved disks write load statistic to use (if installed) iostats, remove hardcoded partitions to support non raspberypi devices
- added php native support for components update with git 
- added cpu info
- added machine brand model and bios description
- show all services states and added ssh
- Added support for running services from PHP when redis not available.
- Move service-runner trigger run responsibility to the model class runService() function.
- User list view return link
- Use model class the right way: remove static usage and global 'nightmare'
- Support disk write load stats without redis
- Preparation for future script execution / components update without redis
- Remove absolute paths from script files (use config base path instead)
- Uniform code across all modules
- Minor UI fixs
- Fix state machine of update logs when needed so that errors can be seen 
- Deal with lack of redis
- Make use of json result codes for all errors across admin views (components, log, firmware and update)
- Move views to ./Views/ folder
- Check symbolic link exists when deleting a module
- Added current user info to main admin page

### CODE REVIEW

**Comment by chaveiro** [comment]
@TrystanLea i consider this work stream finished and all tested, please take a look if some polishing is still needed.
Could you find a way to have the scripts from https://github.com/openenergymonitor/EmonScripts/tree/master/update related to git updates from the adin UI placed at the emoncmc/scripts folder (change repo) so that we dont need to install EmonScripts on shared servers ?

**Comment by chaveiro** [comment]
@TrystanLea  please  take a look at this screenshot, this is how it looks like with your old (and current code).
Is is supposed to look like that on chrome, above the scroll bar ?
image

File: Modules/admin/firmware_view.php [hunk 1]
```diff
@@ -1,17 +1,11 @@
...
-}
-
 #emonhub-running-notice, #emonhub-stopped-notice {
   padding: 8px 8px 8px 14px;
   line-height:31px;
 }
-
 </style>
-
+<div class="admin-container" style="margin-top:10px">
```
Comment (Twincarb) [inline Modules/admin/firmware_view.php:8]: I can't find css class admin-container in a css file but there are several uses where it sets the margin-top:10px its would be good to add the style into admin_styles.css file.

**Comment by chaveiro** [comment]
Got a little carried out on this admin module but now i've finished what i had proposed.
Take a look at the new components manager :)

**Comment by TrystanLea** [comment]
Thanks for this @chaveiro I will take a look soon. Can you do a little peek gif video to demo?: https://github.com/phw/peek

**Comment by chaveiro** [comment]
> Thanks for this @chaveiro I will take a look soon. Can you do a little peek gif video to demo?: https://github.com/phw/peek
Take a look bellow:
Core update
module update
info
overview

**Review by chaveiro — COMMENTED** [review]
An annoying corner case when the window is very small, this https://github.com/emoncms/emoncms/pull/1732/commits/b440acd27b5489256b51a9b82e981c1a0f41b962 makes the menu-l2-controls stick above on bottom left when menu scrollbars appear on the menu-l2-inner.

From this :
image

To this (in red):
 image

File: Modules/admin/admin_model.php [hunk 2]
```diff
@@ -200,202 +266,487 @@ public static function system_information() {
...
+     * @return array
+     */
+    // return the list of available modules minus the installed ones
+    public function components_available() {
+      $localfile = $this->settings['openenergymonitor_dir']."/EmonScripts/components_available.json";
+      $components = array();
+      if (file_exists($localfile)) {
+          $components = json_decode(file_get_contents($localfile), true);
+      }
+      else if ($response = @file_get_contents("https://raw.githubusercontent.com/openenergymonitor/EmonScripts/stable/components_available.json")) {
```
Comment (chaveiro) [inline Modules/admin/admin_model.php:347]: See https://github.com/openenergymonitor/EmonScripts/issues/137

**Comment by chaveiro** [comment]
Hi @TrystanLea could you kindly commit the one, i've moved to fix other parts and don't want to mix up with this as it's open.

**Comment by TrystanLea** [comment]
Hello @chaveiro unfortunately Im a bit swamped at the moment with other things a bad cold and meant to be heading on holiday tomorrow with the family. I would like to study you changes carefully before merging so that I'm confident that I know how it's all working.

Can you continue working on a different branch to keep the commits separate?

Thanks a lot

**Comment by chaveiro** [comment]
I'm really not used to using branches and may end up losing code. For now it's just locally,
FYI have worked on the menu system to fix a lot of issues on corner cases and added sliding. it's just 2 file, i may added it here if you don't mind reviewing it together.
In the pas we had community review from @pb66  and others, not sure how it's now :)
Get better of that cold fast.

**Comment by TrystanLea** [comment]
Using branches is not too difficult, I would at this point just create a new branch for your next stage of development e.g:

    git checkout -b chaveiro_menu_fixes

Then continue with your development and push to your repo copy on github.

This will then keep your pull request on the admin changes above as one stage for me to merge.

Once I merge them, you can then create a pull request from the chaveiro_menu_fixes branch and I can review those changes separately. 

Please do not merge the menu changes into the above changes and as the project maintainer I do need to be the one that reviews these rather than opening it out to a wider group of people. Thanks for understanding.

**Comment by chaveiro** [comment]
Hi, nice to get you testing this. I don't have a raspberry pi at hand so need a little testing from your side to figure it out.
The issue is that apache is running as www-data user (or similar) but the emoncms folder has written permission for user pi only.

Option 1:
Add pi user to the www-data group:
sudo adduser pi www-data

Option 2:
Make apache vhost run as pi user.
Install apache2-mpm-itk.
	apt-get install apache2-mpm-itk
Modify the 000-default.conf (or equivalent) file.
	nano /etc/apache2/sites-available/000-default.conf
Under the DocumentRoot /var/www/html line, add:
	<ifmodule mpm_itk_module>
		AssignUserID pi pi
	</ifmodule>
Restart the apache.
	/etc/init.d/apache2 restart

**Comment by alexandrecuer** [comment]
Install is now working

I add to do : 
```
sudo adduser www-data pi
chmod g+rwx /var/www/emoncms/Modules
```

**Comment by alexandrecuer** [comment]
for components like emonpi, I had to do : 
```
chmod -R g+rwx /opt/openenergymonitor/
```
image

**Comment by alexandrecuer** [comment]
The UI tells that emonhub is not installed : 
image
anyway it is present : 
```
systemctl status emonhub
● emonhub.service - emonHub service description
   Loaded: loaded (/opt/openenergymonitor/emonhub/service/emonhub.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2021-10-05 13:24:57 CEST; 4 days ago
  Process: 397 ExecStartPre=/bin/mkdir -p /var/log/emonhub/ (code=exited, status=0/SUCCESS)
  Process: 428 ExecStartPre=/bin/chgrp -R emonhub /var/log/emonhub/ (code=exited, status=0/SUCCESS)
  Process: 441 ExecStartPre=/bin/chmod 775 /var/log/emonhub/ (code=exited, status=0/SUCCESS)
 Main PID: 447 (python3)
    Tasks: 2 (limit: 2065)
   CGroup: /system.slice/emonhub.service
           └─447 python3 /usr/local/bin/emonhub/emonhub.py --config-file=/etc/emonhub/emonhub.conf --logfile=/var/log/emonhub/emonhub.log

Warning: Journal has been rotated since unit was started. Log output is incomplete or unavailable.
```
**But it is a personal fork**

**Comment by alexandrecuer** [comment]
@chaveiro  : Dont know if all functionnalities are covered, tell me if you want other things to be tested. Tried to update the database from the update view, but it was already OK.

I find the components view a bit slow but it has to investigate the whole system

**Comment by chaveiro** [comment]
Thanks for that. The way this works currently only supports native emoncms modules, the ones that traditionally get installed at /var/www/emoncms/.
To get this working on shared servers, docker and other devices where root is not available or a different instalation path is used, absolute paths like /var/www can't be used.
So all the code does is use the emoncms/Modules folder for reference.
Many modules can just be updated to support this relative paths usage. For specific raspberry pi modules that can't be installed on emoncms/Modules a manual installation is still requires or the process rethinked at the module side.
I've proposed some changes to the available.json here https://github.com/openenergymonitor/EmonScripts/issues/137

**Comment by alexandrecuer** [comment]
read and write load : 
image

machine and CPU
image

**Comment by chaveiro** [comment]
>  e.g Im not sure Im happy yet to merge the changes to the component view to allow the alternative non git approach,

The php approach uses GIT indeed. 
Take a look at the model functions, you will see that the way i architected the code keeps the service runner option first if a script file is available and redis + service runner are running.
So to keep old style approach of using shell scripts to call git instead of php functions, you just need to write some shellscripts for the the new functions that are now done in PHP. 

Now, switch to a high level view and try to reason what might make more sense in the long term after the initial installation of git cloning the core emoncms : 
a - To wrote shell script to call git in a limited way by passing variables there via command line parameters and using an async call via service runner
b - Call an API for doing the same thing and having only to deal with one code base and one programming language for complex validations of the returning data from git in PHP.

From my perspective i think the reasoning should be to get everyting working from the UI in a clean an simple way for the end user to deal with. if it's a raspbery pi, shared server or other.

I understand you option to try to focus on only the product with a raspbery. However currently the effort to support 'all' platforms is easy but if you go the route of raspbery pi only, the community that receive future updated and participate on development will be limited. Similar to excluding redis from new features will leave out a partial number of persons.

> Would you be happy if I go ahead and merge the above partial merges as they are with the reference to your pull request..
Sure i don't mind about the authorship of this, i just want the code to get on main so i can sync back and continue :)

**Comment by TrystanLea** [comment]
Hello @chaveiro would you be willing to help me merge the above changes by separating out this pull request into smaller stages. I think the easiest way for you to do this would be to use your Beyond Compare tool to create new pull requests that contain one specific set of changes at a time. 

A starting example would be a pull request specifically for:

`Make use of json result codes for all errors across admin views (components, log, firmware and update)`

There's a fair bit of code Id be happy to merge quite quickly if broken down into smaller parts.

I came across this guide for the kubernetes project the other day and it describes really well what I'm after:
 https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#smaller-is-better-small-commits-small-pull-requests

I do appreciate this is more time for you.

**Comment by chaveiro** [comment]
Here we go again.. :)
I pushed individually commits for this with that in mind. If you can follow the commits history it gets easier to understand.
Let me see If i can take a look during the next week, this is full of conflicts now and I'm not sure why.
Please try to look at the code as a whole and do a review. This is very straightforward.

**Comment by TrystanLea** [comment]
Thanks @chaveiro 

I fixed the conflicts in this branch https://github.com/emoncms/emoncms/compare/chaveiro_admin_development...chaveiro:master you should be able to pull that into your master branch hopefully and then re-create a pull request to the emoncms master.. 

I think my request is reasonable, see the kubernetes discussion on this. I could review as a whole but it would take me a good number of days work to do that and it means if there is any part I don't want to merge it will hold up all the other changes. 

Thanks for offering to take a look at it

**Comment by TrystanLea** [comment]
The other option is that I try and go through each commit using git cherry-pick I started trying that way last night: https://github.com/emoncms/emoncms/commit/d954f8f9bf3d9ebdfa221806c7a48bf197e0aeed but not sure if that's going to create more problems if I make small modifications along the way... I can try it if your then happy to help me with conflicts that might arrise

**Comment by TrystanLea** [comment]
Using the git cherry-pick tool, I could merge:

- https://github.com/emoncms/emoncms/commit/285499bc2f5fd315cbdb8a26d12bc7059d75d617
- https://github.com/emoncms/emoncms/commit/4c54a8c3fe35751cb03dc811d6daaeb9c063838a

Then there's too much happening on this one:

- https://github.com/emoncms/emoncms/commit/d337f80df71c5b4d26b3bf8c51061bac366e065c

The following one has a lot happening but I think I could work my way through that one and merge:

https://github.com/emoncms/emoncms/commit/54156e5ddfa446d3f30196e49773a8b988dd39ac

Not sure what happens if I skip d337f80df71c5b4d26b3bf8c51061bac366e065c and merge 54156e5ddfa446d3f30196e49773a8b988dd39ac feels like things could get messy.

The first two commits are great, small and concise, easy to merge, it then gets a lot harder and I struggle to follow

**Comment by TrystanLea** [comment]
Which is why I think it might work better if you create new pull requests that pull out specific changes one at a time. Each pull request would do one thing at a time e.g *Make use of json result codes for all errors across admin views (components, log, firmware and update)* and would not contain any other changes that do not relate to that change e.g: refactoring, css/layout changes etc.

**Comment by TrystanLea** [comment]
Ok d337f80 is a git chery-pick skippable commit. So lets see if I can do the next

**Comment by TrystanLea** [comment]
Ok: 

- https://github.com/emoncms/emoncms/commit/fbea39126f4b41c83c18170f8c65da69e3bb0990
- https://github.com/emoncms/emoncms/commit/55fa35dbc6c3ab7d4dded7a5e8adb15e5009c44c
- https://github.com/emoncms/emoncms/commit/913629e99a6aab563f4abe3ccc9cbbc257a07cba

have now been merged,

The branch that I mentioned above that I had fixed conflicts in is now full of conflicts again when comparing with master so may be best to ignore that branch now, will keep going with cherry-pick see were I get too

**Comment by TrystanLea** [comment]
A couple more commits merged to master:

- https://github.com/emoncms/emoncms/commit/67fc05a2eadd9626471fcd6a55492f5721be6723
- https://github.com/emoncms/emoncms/commit/ae755ebfe55db4cebe923743214086c6473a5258
- https://github.com/emoncms/emoncms/commit/e20fb5901d02084e11d055b6af80f729b57a7e8c
- https://github.com/emoncms/emoncms/commit/3dd37aa5d395b0f3e38a379bda90d21c8ac4cdc4

**Comment by TrystanLea** [comment]
And some more merged:

- https://github.com/emoncms/emoncms/commit/3a506434ec89adbcc418873a7c80c008edb5600d
- https://github.com/emoncms/emoncms/commit/f232233ba1d77012a005a87cec2c49275ff371b8
- https://github.com/emoncms/emoncms/commit/8befed90d4a0e575c383c45c4ffd0dace45128e8
- https://github.com/emoncms/emoncms/commit/a9bd2e895720400b3d6c20c5d73b285921364ab9
- https://github.com/emoncms/emoncms/commit/3293bd520c7ee6da9e1c38990f6a3621d23ebbd0
- https://github.com/emoncms/emoncms/commit/a7ed96fde1608e5359b5d5b25f4140643c9f4ffb
- https://github.com/emoncms/emoncms/commit/813a85c84ca5e1aab608294ec8ad38bb0e7aefdb

I will keep on with this approach, will let you know when I have finished and if I need further input at that stage.

**Comment by TrystanLea** [comment]
More commits merged: 

- https://github.com/emoncms/emoncms/commit/813a85c84ca5e1aab608294ec8ad38bb0e7aefdb
- https://github.com/emoncms/emoncms/commit/484f936fd4719945883b938006277849218e10e2
- https://github.com/emoncms/emoncms/commit/cd71480f8a45bf46cbd109054b539b42b0650a5a
- https://github.com/emoncms/emoncms/commit/dc2fbab1d11c76541054b2a2c2343231b58f8e59
- https://github.com/emoncms/emoncms/commit/909dca4da63ed4bb85bd32f893bec1141c67a524

The remaining commits are all to do with the component manager which I would like to review more carefully.
I will try and come back to that when I next get some time hopefully later next week.

Not sure if its worth you rebasing your master branch on main now or waiting until I've reviewed the remaining commits.

**Comment by chaveiro** [comment]
Hi @trystan, I've tried to look at this and got a little confused, what is currently the status.
May i suggest to do it like previously you do a code review and i answer and change behaviors to your preference?

**Comment by TrystanLea** [comment]
Sure, I've merged most of the commits up to those affecting the component manager. I need to get through some other work before I get back to this, so it will probably be a couple of weeks before I can continue the review. 

If you are ok to wait with this pull request linked to your master branch then feel free to do that, but if you did want to free up your master branch to pull in upstream changes, then it might be worth creating a new branch with the remaining commits that are yet to be merged inside. 

If you create two branches one that just holds your current master branch status e.g:

    git checkout -b master_backup

Then do another to test merging in the current upstream master:

    git checkout -b master_merge_test
    git fetch upstream
    git merge upstream:master

If you think you can fix the conflicts that result then give that a go. If you get stuck you can always delete your merge test e.g:

    git checkout master_backup
    git branch -D master_merge_test

If you do manage to fix the conflicts it might be worth trying to open a pull request from your master_merge_test branch against the upstream emoncms master.

In hindsight I should have avoided the initial manual merge that I did moving the view file locations to make the git diff easier and just followed the git cherry-pick process above from the start. I will try that first next time.

**Comment by TrystanLea** [comment]
Hello @chaveiro 

I've fixed the merge conflicts resulting from my partial merge and pushed the result to a new branch here https://github.com/emoncms/emoncms/pull/1765 from which I will continue the review process. You should be able to pull in from that branch to updated your master branch to include all the other recent commits to emoncms master if you want. Il create a pull request to your master branch.

Closing this pull request to work from #1765

Thanks

**Pull request closed without merging**